tenantcostserver: use the tenant_usage system table

RaduBerinde opened the pull request:
This PR prototypes a new tenant_usage schema  based on @ajwerner's suggestion in the RFC (#66436). I will update the RFC after getting some initial feedback here.

#### sql: add tenant_usage system table

Add the system table described in the RFC (#66436).

The table is only created for the system tenant.

Release note: None

#### tenantcostserver: use the tenant_usage system table

This change implements most of the interaction with the tenant_usage
system table, with the exception of dead instances detection and
clean-up.

We currently tolerate an empty table, but it would be cleaner to
initialize the tenant state (instance_id=0 row) at tenant creation
time (+ implement a migration). I will explore this in a future
change, when we add some configurable defaults for the refill rate
etc.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/68115)
<!-- Reviewable:end -->

ajwerner reviewed (commented):
I like it. How do you feel about it?

> Reviewed 32 of 32 files at r1, 6 of 11 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68115#-:-MfeZDQ_6BROBsFqsKdQ:b-si8h8d)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru and @RaduBerinde)

___
*[pkg/ccl/multitenantccl/tenantcostserver/system_table.go, line 371 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/68115#-MfeXAQeEQ2rnw2fai0a:-MfeXAQeEQ2rnw2fai0b:b-xmfcji) ([raw file](https://github.com/cockroachdb/cockroach/blob/f4f0be388a7efcefe7b00788514316903a0cb283/pkg/ccl/multitenantccl/tenantcostserver/system_table.go#L371)):*
> ```Go
> // checkInvariants reads all rows in the system table for the given tenant and
> // checks that the state is consistent.
> func (h *sysTableHelper) checkInvariants() error {
> ```

I find it surprising that this method returns an error but also panics. Did you intend to convert those to errors?
___
*[pkg/ccl/multitenantccl/tenantcostserver/system_table.go, line 381 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/68115#-MfeWgQl3ia94cuq6dqD:-MfeWgQl3ia94cuq6dqE:b88ta8c) ([raw file](https://github.com/cockroachdb/cockroach/blob/f4f0be388a7efcefe7b00788514316903a0cb283/pkg/ccl/multitenantccl/tenantcostserver/system_table.go#L381)):*
> ```Go
> 			next_instance_id,          /* 1 */
> 			ru_burst_limit,            /* 2 */
> 			ru_refill_rate,            /* 3 */        
> ```

trailing spaces
___
*[pkg/ccl/multitenantccl/tenantcostserver/token_bucket.go, line 35 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/68115#-MfeYD8R0D5IG3gUGBrV:-MfeYD8R0D5IG3gUGBrW:b9gkbb2) ([raw file](https://github.com/cockroachdb/cockroach/blob/f4f0be388a7efcefe7b00788514316903a0cb283/pkg/ccl/multitenantccl/tenantcostserver/token_bucket.go#L35)):*
> ```Go
> 	}
> 
> 	result := &roachpb.TokenBucketResponse{}
> ```

consider resetting the result at the top of the closure. If you make sure to update every field on all call paths, it's fine, but I've been bitten by early returns and populated results in the past. 
___
*[pkg/ccl/multitenantccl/tenantcostserver/token_bucket.go, line 74 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/68115#-MfeYXPH2ufzUr9GYj-e:-MfeYXPH2ufzUr9GYj-f:b-3ecgh5) ([raw file](https://github.com/cockroachdb/cockroach/blob/f4f0be388a7efcefe7b00788514316903a0cb283/pkg/ccl/multitenantccl/tenantcostserver/token_bucket.go#L74)):*
<details><summary><i alt="summary">Quoted 7 lines of code…</i></summary>

> ```
> 		m := s.metrics.getTenantMetrics(tenantID)
> 		m.totalRU.Update(tenant.Consumption.RU)
> 		m.totalReadRequests.Update(int64(tenant.Consumption.ReadRequests))
> 		m.totalReadBytes.Update(int64(tenant.Consumption.ReadBytes))
> 		m.totalWriteRequests.Update(int64(tenant.Consumption.WriteRequests))
> 		m.totalWriteBytes.Update(int64(tenant.Consumption.WriteBytes))
> 		m.totalSQLPodsCPUSeconds.Update(tenant.Consumption.SQLPodCPUSeconds)
> ```
</details>

Should this happen outside of the closure?
___
*[pkg/sql/catalog/systemschema/system.go, line 515 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68115#-MfeVbaM3LTvXU8u0Zor:-MfeVbaM3LTvXU8u0Zos:b9ogee2) ([raw file](https://github.com/cockroachdb/cockroach/blob/fb798012165736d1590b187e31d4b1b5b5b6fe3b/pkg/sql/catalog/systemschema/system.go#L515)):*
<details><summary><i alt="summary">Quoted 50 lines of code…</i></summary>

> ```
>   tenant_id INT NOT NULL,
> 	-- For each tenant, there is a special row with instance_id = 0 which contains
> 	-- per-tenant stat. Each SQL instance (pod) also has its own row with
> 	-- per-instance state.
> 	instance_id INT NOT NULL,
> 	-- next_instance_id identifies the next live instance ID, with the smallest ID
> 	-- larger than this instance_id (or 0 if there is no such ID).
> 	-- We are overlaying a circular linked list of all live instances, with
> 	-- instance 0 acting as a sentinel (always the head of the list).
> 	next_instance_id INT NOT NULL,
> 	-- -------------------------------------------------------------------
> 	--  The following fields are used only for the per-tenant state, when
> 	--  instance_id = 0.
> 	-- -------------------------------------------------------------------
>   -- Bucket configuration.
>   ru_burst_limit FLOAT,
>   ru_refill_rate FLOAT,
>   -- Current amount of RUs in the bucket.
>   ru_current FLOAT,
>   -- Current sum of the shares values for all instances.
>   current_share_sum FLOAT,
>   -- Cumulative usage statistics.
>   total_ru_usage            FLOAT,
>   total_read_requests       INT,
>   total_read_bytes          INT,
>   total_write_requests      INT,
>   total_write_bytes         INT,
>   total_sql_pod_cpu_seconds FLOAT, -- TODO: Maybe milliseconds and INT8?
> 	-- -------------------------------------------------------------
> 	--  The following fields are used for per-instance state, when
> 	--  instance_id != 0.
> 	-- --------------------------------------------------------------
> 	-- The lease is a unique identifier for this instance, necessary because
> 	-- instance IDs can be reused.
> 	instance_lease BYTES,
> 	-- Last request sequence number. These numbers are provided by the
> 	-- instance and are monotonically increasing; used to detect duplicate
> 	-- requests and provide idempotency.
> 	instance_seq INT,
> 	-- Current shares value for this instance.
>   instance_shares FLOAT,
> 	-- Time when we last heard from this instance.
> 	instance_last_update TIMESTAMP,
> 	FAMILY "primary" (
> 	  tenant_id, instance_id, next_instance_id,
> 		ru_burst_limit, ru_refill_rate, ru_current, current_share_sum,
> 		total_ru_usage, total_read_requests, total_read_bytes, total_write_requests,
> 		total_write_bytes, total_sql_pod_cpu_seconds,
> 		instance_lease, instance_seq, instance_shares, instance_last_update
> 	),
> ```
</details>

nit: the indentation is a bit off
___
*[pkg/sql/tests/system_table_test.go, line 215 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68115#-MfeW-hb-4oFqA2wSRfY:-MfeW-hc8axyvetU8Atj:b8dsdmr) ([raw file](https://github.com/cockroachdb/cockroach/blob/fb798012165736d1590b187e31d4b1b5b5b6fe3b/pkg/sql/tests/system_table_test.go#L215)):*
> ```
> 			t.Errorf("\n%#v\n", test.pkg.TableDesc())
> 			t.Errorf("\n%#v\n", gen.TableDesc())
> ```
nit: either remove or move into the below error call


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
I like it too. Having the per-instance state makes things less fragile and easier to debug/understand, and it is more extensible - I can imagine various features that would need per-instance state.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68115#-:-Mfio27WCtQ8YZTJezsg:b-sbngx0)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru and @ajwerner)

___
*[pkg/ccl/multitenantccl/tenantcostserver/system_table.go, line 371 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/68115#-MfeXAQeEQ2rnw2fai0a:-Mfip55H339iS4NpMcNr:b-o419a3) ([raw file](https://github.com/cockroachdb/cockroach/blob/f4f0be388a7efcefe7b00788514316903a0cb283/pkg/ccl/multitenantccl/tenantcostserver/system_table.go#L371)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

I find it surprising that this method returns an error but also panics. Did you intend to convert those to errors?
</blockquote></details>

Right, fixed.
___
*[pkg/ccl/multitenantccl/tenantcostserver/system_table.go, line 381 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/68115#-MfeWgQl3ia94cuq6dqD:-MfipCCM3lVBZHml86l8:bt2vr1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/f4f0be388a7efcefe7b00788514316903a0cb283/pkg/ccl/multitenantccl/tenantcostserver/system_table.go#L381)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

trailing spaces
</blockquote></details>

Fixed.
___
*[pkg/ccl/multitenantccl/tenantcostserver/token_bucket.go, line 35 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/68115#-MfeYD8R0D5IG3gUGBrV:-MfipM0A5MuNmgrMsy-5:bohqzsr) ([raw file](https://github.com/cockroachdb/cockroach/blob/f4f0be388a7efcefe7b00788514316903a0cb283/pkg/ccl/multitenantccl/tenantcostserver/token_bucket.go#L35)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

consider resetting the result at the top of the closure. If you make sure to update every field on all call paths, it's fine, but I've been bitten by early returns and populated results in the past. 
</blockquote></details>

Agreed, done.
___
*[pkg/ccl/multitenantccl/tenantcostserver/token_bucket.go, line 74 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/68115#-MfeYXPH2ufzUr9GYj-e:-MfipO0M8d_FjZFeaIzx:bjzuq5t) ([raw file](https://github.com/cockroachdb/cockroach/blob/f4f0be388a7efcefe7b00788514316903a0cb283/pkg/ccl/multitenantccl/tenantcostserver/token_bucket.go#L74)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

> ```
> 		m := s.metrics.getTenantMetrics(tenantID)
> 		m.totalRU.Update(tenant.Consumption.RU)
> 		m.totalReadRequests.Update(int64(tenant.Consumption.ReadRequests))
> 		m.totalReadBytes.Update(int64(tenant.Consumption.ReadBytes))
> 		m.totalWriteRequests.Update(int64(tenant.Consumption.WriteRequests))
> 		m.totalWriteBytes.Update(int64(tenant.Consumption.WriteBytes))
> 		m.totalSQLPodsCPUSeconds.Update(tenant.Consumption.SQLPodCPUSeconds)
> ```
Should this happen outside of the closure?
</blockquote></details>

I was concerned with having a race between two different requests and having them update the metrics in opposite order. But this has a different problem, we may update the metrics even if our transaction ultimately fails to commit.

I moved them out and added a TODO. Will go away if we implement the rangefeeds solution.


<!-- Sent from Reviewable.io -->

andy-kimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68115)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @ajwerner, and @RaduBerinde)

___
*[pkg/ccl/multitenantccl/tenantcostserver/system_table.go, line 31 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/68115#-Mfrj_lM8aAUQeco8WZ-:-Mfrj_lM8aAUQeco8WZ0:b-3hevp8) ([raw file](https://github.com/cockroachdb/cockroach/blob/a86f7187a3c2dfc36da3fec6550865aa4283d6dd/pkg/ccl/multitenantccl/tenantcostserver/system_table.go#L31)):*
> ```Go
> // pod for this tenant has a different instance ID (but note that IDs can be
> // reused).
> type InstanceID int32
> ```

Are you avoiding importing  `base.SQLInstanceID`?
___
*[pkg/ccl/multitenantccl/tenantcostserver/system_table.go, line 67 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/68115#-MfrlzZa801tUa_cw03i:-MfrlzZa801tUa_cw03j:b-sg76gt) ([raw file](https://github.com/cockroachdb/cockroach/blob/a86f7187a3c2dfc36da3fec6550865aa4283d6dd/pkg/ccl/multitenantccl/tenantcostserver/system_table.go#L67)):*
> ```Go
> 	// incarnations of the same ID.
> 	Lease tree.DBytes
> 	// Seq is a sequence number used to
> ```

Comment cut off.
___
*[pkg/ccl/multitenantccl/tenantcostserver/system_table.go, line 282 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/68115#-Mfrou-LBWPD9yu6KfJR:-Mfrou-LBWPD9yu6KfJS:b-2xw1af) ([raw file](https://github.com/cockroachdb/cockroach/blob/a86f7187a3c2dfc36da3fec6550865aa4283d6dd/pkg/ccl/multitenantccl/tenantcostserver/system_table.go#L282)):*
> ```Go
> // next_instance_id in the table (or updates tenant.FirstInstance).
> //
> // Note that this should only happen when a SQL pod starts up (it is not in the
> ```

We do consider startup of a SQL pod as a "hot path" and we carefully measure/monitor this time. This is because the latency of resuming from suspension is so important. If the tenant system table is in another region from the SQL pod, we'll want to make sure that creation of the pod isn't delayed by a roundtrip call to the table. From what I can see, we're only calling this lazily when we discover an instance does not yet exist, as part of reporting tenant usage to the server. That's good, since that will definitely keep this out of the pod startup hot path. But you may want to tweak this comment to make it more clear this won't delay pod startup.
___
*[pkg/ccl/multitenantccl/tenantcostserver/system_table.go, line 286 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/68115#-MfrsH2a97GC8jTimk3a:-MfrsH2a97GC8jTimk3b:be0d095) ([raw file](https://github.com/cockroachdb/cockroach/blob/a86f7187a3c2dfc36da3fec6550865aa4283d6dd/pkg/ccl/multitenantccl/tenantcostserver/system_table.go#L286)):*
> ```Go
> func (h *sysTableHelper) accomodateNewInstance(tenant *tenantState, instance *instanceState) error {
> 	if tenant.FirstInstance == 0 || tenant.FirstInstance > instance.ID {
> 		// Thew new instance has the lowest ID.
> ```

SP: The
___
*[pkg/ccl/multitenantccl/tenantcostserver/system_table.go, line 328 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/68115#-Mfrt8etANyrjV6GB-lW:-Mfrt8etANyrjV6GB-lX:bxo0hvl) ([raw file](https://github.com/cockroachdb/cockroach/blob/a86f7187a3c2dfc36da3fec6550865aa4283d6dd/pkg/ccl/multitenantccl/tenantcostserver/system_table.go#L328)):*
> ```Go
> 		h.ctx, "update-next-id", h.txn,
> 		sessiondata.NodeUserSessionDataOverride,
> 		// Update the previous instance's next_instnace_id.
> ```

SP: next_instance_id
___
*[pkg/ccl/multitenantccl/tenantcostserver/token_bucket.go, line 82 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/68115#-Mfs6pYNDlnGaaq8oUCL:-Mfs5aau-N_XuZFjQdzs:b-2p7lbq) ([raw file](https://github.com/cockroachdb/cockroach/blob/a86f7187a3c2dfc36da3fec6550865aa4283d6dd/pkg/ccl/multitenantccl/tenantcostserver/token_bucket.go#L82)):*
> ```Go
> 
> 	// Report current consumption.
> 	// TODO(radu): there is a possible race here, where two different requests
> ```

@ajwerner, how come these metrics are gauges and not counters? They can only go up, right?

To fix the race, can we add a `CompareAndSwap` primitive to the gauge struct, or maybe a `SetIfHigher` primitive to the counter struct? In both cases, we'd ensure that these metrics can only go up.


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68115)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @ajwerner, @andy-kimball, and @RaduBerinde)

___
*[pkg/ccl/multitenantccl/tenantcostserver/token_bucket.go, line 82 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/68115#-Mfs6pYNDlnGaaq8oUCL:-MfsD4_K6rG0Pi41PqH5:b13epe8) ([raw file](https://github.com/cockroachdb/cockroach/blob/a86f7187a3c2dfc36da3fec6550865aa4283d6dd/pkg/ccl/multitenantccl/tenantcostserver/token_bucket.go#L82)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

@ajwerner, how come these metrics are gauges and not counters? They can only go up, right?

To fix the race, can we add a `CompareAndSwap` primitive to the gauge struct, or maybe a `SetIfHigher` primitive to the counter struct? In both cases, we'd ensure that these metrics can only go up.
</blockquote></details>

I don't think they can be counters in that we want them to reflect the aggregate value across all instances as observed from any kv node and the nodes communicate with each other through the table. I'm not clear on how we could use a counter. 

A `SetIfHigher` primitive sounds reasonable and straightforward. 


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68115)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @ajwerner, @andy-kimball, and @RaduBerinde)

___
*[pkg/ccl/multitenantccl/tenantcostserver/token_bucket.go, line 82 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/68115#-Mfs6pYNDlnGaaq8oUCL:-MfsEsjP0joxEyWgvl2t:bga2ati) ([raw file](https://github.com/cockroachdb/cockroach/blob/a86f7187a3c2dfc36da3fec6550865aa4283d6dd/pkg/ccl/multitenantccl/tenantcostserver/token_bucket.go#L82)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

I don't think they can be counters in that we want them to reflect the aggregate value across all instances as observed from any kv node and the nodes communicate with each other through the table. I'm not clear on how we could use a counter. 

A `SetIfHigher` primitive sounds reasonable and straightforward. 
</blockquote></details>

Speaking of the race, should we have a mutex or something to only allow one request for a tenant to proceed on a given gateway at a time? It feels like we can avoid extra contention here and also stop thinking about this race. 


<!-- Sent from Reviewable.io -->

andy-kimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68115)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @ajwerner, and @RaduBerinde)

___
*[pkg/ccl/multitenantccl/tenantcostserver/token_bucket.go, line 82 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/68115#-Mfs6pYNDlnGaaq8oUCL:-MfsEKyBCSkUt3oVtZxW:b-e6f9hl) ([raw file](https://github.com/cockroachdb/cockroach/blob/a86f7187a3c2dfc36da3fec6550865aa4283d6dd/pkg/ccl/multitenantccl/tenantcostserver/token_bucket.go#L82)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Speaking of the race, should we have a mutex or something to only allow one request for a tenant to proceed on a given gateway at a time? It feels like we can avoid extra contention here and also stop thinking about this race. 
</blockquote></details>

From the [Prometheus documentation](https://prometheus.io/docs/practices/instrumentation/#counter-vs-gauge-summary-vs-histogram):
```
Counter vs. gauge, summary vs. histogram
It is important to know which of the four main metric types to use for a given metric.

To pick between counter and gauge, there is a simple rule of thumb: if the value can go down, it is a gauge.

Counters can only go up (and reset, such as when a process restarts). They are useful for accumulating the number of events, or the amount of something at each event. For example, the total number of HTTP requests, or the total number of bytes sent in HTTP requests. Raw counters are rarely useful. Use the rate() function to get the per-second rate at which they are increasing.

Gauges can be set, go up, and go down. They are useful for snapshots of state, such as in-progress requests, free/total memory, or temperature. You should never take a rate() of a gauge.
```
If Prometheus knows that a value can only go up (like these ones), it can do a better job handling gaps in the data as well as resets to zero. I realize that we currently only have an `Inc` method on the counter class, but I don't see a reason why a `UpdateIfHigher` method wouldn't also fit the intent and constraints of a counter. As long as the value always goes up, it doesn't seem important how it gets there - through delta increment or absolute update.


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68115)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @ajwerner, @andy-kimball, and @RaduBerinde)

___
*[pkg/ccl/multitenantccl/tenantcostserver/token_bucket.go, line 82 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/68115#-Mfs6pYNDlnGaaq8oUCL:-MfsRUGy6dCH5r4mhwNK:brd71xr) ([raw file](https://github.com/cockroachdb/cockroach/blob/a86f7187a3c2dfc36da3fec6550865aa4283d6dd/pkg/ccl/multitenantccl/tenantcostserver/token_bucket.go#L82)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

From the [Prometheus documentation](https://prometheus.io/docs/practices/instrumentation/#counter-vs-gauge-summary-vs-histogram):
```
Counter vs. gauge, summary vs. histogram
It is important to know which of the four main metric types to use for a given metric.

To pick between counter and gauge, there is a simple rule of thumb: if the value can go down, it is a gauge.

Counters can only go up (and reset, such as when a process restarts). They are useful for accumulating the number of events, or the amount of something at each event. For example, the total number of HTTP requests, or the total number of bytes sent in HTTP requests. Raw counters are rarely useful. Use the rate() function to get the per-second rate at which they are increasing.

Gauges can be set, go up, and go down. They are useful for snapshots of state, such as in-progress requests, free/total memory, or temperature. You should never take a rate() of a gauge.
```
If Prometheus knows that a value can only go up (like these ones), it can do a better job handling gaps in the data as well as resets to zero. I realize that we currently only have an `Inc` method on the counter class, but I don't see a reason why a `UpdateIfHigher` method wouldn't also fit the intent and constraints of a counter. As long as the value always goes up, it doesn't seem important how it gets there - through delta increment or absolute update.
</blockquote></details>

Makes sense. We can add an `UpdateIfHigher` method to the counter type with equal ease as to the gauge. 


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68115)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @ajwerner, @andy-kimball, and @RaduBerinde)

___
*[pkg/ccl/multitenantccl/tenantcostserver/token_bucket.go, line 82 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/68115#-Mfs6pYNDlnGaaq8oUCL:-MfsgM0aAEH8brH2owIS:bvodeci) ([raw file](https://github.com/cockroachdb/cockroach/blob/a86f7187a3c2dfc36da3fec6550865aa4283d6dd/pkg/ccl/multitenantccl/tenantcostserver/token_bucket.go#L82)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Makes sense. We can add an `UpdateIfHigher` method to the counter type with equal ease as to the gauge. 
</blockquote></details>

Part of the problem is that a couple of these are floats (including the most important one, the RUs). We don't have a float counter. We could add it but I don't know to what extent a float counter would make sense to Prometheus?

A mutex SGTM.


<!-- Sent from Reviewable.io -->

andy-kimball commented:
___
*[pkg/ccl/multitenantccl/tenantcostserver/token_bucket.go, line 82 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/68115#-Mfs6pYNDlnGaaq8oUCL:-Mfsk7eK0Ri1pSUqLCTE:b2h3cyl) ([raw file](https://github.com/cockroachdb/cockroach/blob/a86f7187a3c2dfc36da3fec6550865aa4283d6dd/pkg/ccl/multitenantccl/tenantcostserver/token_bucket.go#L82)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Part of the problem is that a couple of these are floats (including the most important one, the RUs). We don't have a float counter. We could add it but I don't know to what extent a float counter would make sense to Prometheus?

A mutex SGTM.
</blockquote></details>

The Prometheus Go library returns all samples as `float64`:
```
// A SampleValue is a representation of a value for a given sample at a given
// time.
type SampleValue float64
```


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68115)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @ajwerner, @andy-kimball, and @RaduBerinde)

___
*[pkg/ccl/multitenantccl/tenantcostserver/token_bucket.go, line 82 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/68115#-Mfs6pYNDlnGaaq8oUCL:-MfslZWdD6QqFnvFT7Nv:b7v7ck6) ([raw file](https://github.com/cockroachdb/cockroach/blob/a86f7187a3c2dfc36da3fec6550865aa4283d6dd/pkg/ccl/multitenantccl/tenantcostserver/token_bucket.go#L82)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

The Prometheus Go library returns all samples as `float64`:
```
// A SampleValue is a representation of a value for a given sample at a given
// time.
type SampleValue float64
```
</blockquote></details>

I see. I will investigate switching to counters in a separate change.


<!-- Sent from Reviewable.io -->

andy-kimball commented:
___
*[pkg/ccl/multitenantccl/tenantcostserver/token_bucket.go, line 82 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/68115#-Mfs6pYNDlnGaaq8oUCL:-MfsxDtlAxFKV1-lUtBw:blkn3dh) ([raw file](https://github.com/cockroachdb/cockroach/blob/a86f7187a3c2dfc36da3fec6550865aa4283d6dd/pkg/ccl/multitenantccl/tenantcostserver/token_bucket.go#L82)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

I see. I will investigate switching to counters in a separate change.
</blockquote></details>

Maybe open a task with this info. This could be a task 


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68115)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @ajwerner, and @andy-kimball)

___
*[pkg/ccl/multitenantccl/tenantcostserver/system_table.go, line 31 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/68115#-Mfrj_lM8aAUQeco8WZ-:-MfsyMrO9Ip35uaXMk2c:bt2vr1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/a86f7187a3c2dfc36da3fec6550865aa4283d6dd/pkg/ccl/multitenantccl/tenantcostserver/system_table.go#L31)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Are you avoiding importing  `base.SQLInstanceID`?
</blockquote></details>

Fixed.
___
*[pkg/ccl/multitenantccl/tenantcostserver/system_table.go, line 67 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/68115#-MfrlzZa801tUa_cw03i:-MfsyXdG5indFzp0cYCf:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a86f7187a3c2dfc36da3fec6550865aa4283d6dd/pkg/ccl/multitenantccl/tenantcostserver/system_table.go#L67)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Comment cut off.
</blockquote></details>

Done.
___
*[pkg/ccl/multitenantccl/tenantcostserver/system_table.go, line 282 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/68115#-Mfrou-LBWPD9yu6KfJR:-MfsyryjC1zCtKD4oqIO:b-1amf0j) ([raw file](https://github.com/cockroachdb/cockroach/blob/a86f7187a3c2dfc36da3fec6550865aa4283d6dd/pkg/ccl/multitenantccl/tenantcostserver/system_table.go#L282)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

We do consider startup of a SQL pod as a "hot path" and we carefully measure/monitor this time. This is because the latency of resuming from suspension is so important. If the tenant system table is in another region from the SQL pod, we'll want to make sure that creation of the pod isn't delayed by a roundtrip call to the table. From what I can see, we're only calling this lazily when we discover an instance does not yet exist, as part of reporting tenant usage to the server. That's good, since that will definitely keep this out of the pod startup hot path. But you may want to tweak this comment to make it more clear this won't delay pod startup.
</blockquote></details>

Yes, we are on the same page, improved the comment.
___
*[pkg/ccl/multitenantccl/tenantcostserver/system_table.go, line 286 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/68115#-MfrsH2a97GC8jTimk3a:-Mfsyw5P7uQw_CRDbntW:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a86f7187a3c2dfc36da3fec6550865aa4283d6dd/pkg/ccl/multitenantccl/tenantcostserver/system_table.go#L286)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

SP: The
</blockquote></details>

Done.
___
*[pkg/ccl/multitenantccl/tenantcostserver/system_table.go, line 328 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/68115#-Mfrt8etANyrjV6GB-lW:-MfsyySs68f7g0NM0Hxu:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a86f7187a3c2dfc36da3fec6550865aa4283d6dd/pkg/ccl/multitenantccl/tenantcostserver/system_table.go#L328)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

SP: next_instance_id
</blockquote></details>

Done.
___
*[pkg/ccl/multitenantccl/tenantcostserver/token_bucket.go, line 82 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/68115#-Mfs6pYNDlnGaaq8oUCL:-MfszXvk2cRvo-zPRCuN:b-satkk1) ([raw file](https://github.com/cockroachdb/cockroach/blob/a86f7187a3c2dfc36da3fec6550865aa4283d6dd/pkg/ccl/multitenantccl/tenantcostserver/token_bucket.go#L82)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Maybe open a task with this info. This could be a task 
</blockquote></details>

Filed https://github.com/cockroachdb/cockroach/issues/68291

Added the mutex in separate commit.


<!-- Sent from Reviewable.io -->

RaduBerinde committed:
sql: add tenant_usage system table

Add the system table described in the RFC (#66436).

The table is only created for the system tenant.

Release note: None

RaduBerinde committed:
tenantcostserver: use a per-tenant mutex

Use a per-tenant mutex to avoid unnecessary transaction restart and
remove the possibility of out-of-order metric updates.

Release note: None

RaduBerinde committed:
tenantcostserver: use the tenant_usage system table

This change implements most of the interaction with the tenant_usage
system table, with the exception of dead instances detection and
clean-up.

We currently tolerate an empty table, but it would be cleaner to
initialize the tenant state (instance_id=0 row) at tenant creation
time (+ implement a migration). I will explore this in a future
change, when we add some configurable defaults for the refill rate
etc.

Release note: None

andy-kimball reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>, with a few nits.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68115#-:-Mftb9m54yfvkeHZwZ1X:b-rgpnmw)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @adityamaru, @ajwerner, @andy-kimball, and @RaduBerinde)

___
*[pkg/ccl/multitenantccl/tenantcostserver/metrics.go, line 112 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/68115#-Mft_uNSAVVC_NwmsZC2:-Mft_uNSAVVC_NwmsZC3:b-cgaarg) ([raw file](https://github.com/cockroachdb/cockroach/blob/0f9241607580ce768750cfdd2d5c5c646e6f8dcb/pkg/ccl/multitenantccl/tenantcostserver/metrics.go#L112)):*
> ```Go
> 	// Mutex is used to atomically update metrics together with a corresponding
> 	// change to the system table.
> 	mutex *syncutil.Mutex
> ```

I don't think this needs to be a pointer, does it? Why allocate an extra object?
___
*[pkg/ccl/multitenantccl/tenantcostserver/system_table.go, line 33 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/68115#-MftaAH59cxD7DDdHgXq:-MftaAH59cxD7DDdHgXr:b-curzcq) ([raw file](https://github.com/cockroachdb/cockroach/blob/0f9241607580ce768750cfdd2d5c5c646e6f8dcb/pkg/ccl/multitenantccl/tenantcostserver/system_table.go#L33)):*
> ```Go
> 	Present bool
> 
> 	// Smallest active instance ID.
> ```

SUPER NIT: These comments should start with field name to be idiomatic go; similar issues in `instanceState`.
___
*[pkg/ccl/multitenantccl/tenantcostserver/token_bucket.go, line 82 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/68115#-Mfs6pYNDlnGaaq8oUCL:-Mft_PzbCmvDs9sO0pw0:b-3odv2r) ([raw file](https://github.com/cockroachdb/cockroach/blob/a86f7187a3c2dfc36da3fec6550865aa4283d6dd/pkg/ccl/multitenantccl/tenantcostserver/token_bucket.go#L82)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Filed https://github.com/cockroachdb/cockroach/issues/68291

Added the mutex in separate commit.
</blockquote></details>

Whoops, my comment got cut off. I meant to say, "This could be task for someone else". Also, you could add the work to fix the TODO in that task by adding an `UpdateIfHigher` or `RatchetUp` method. How come you removed the TODO comment from the metrics update code below?
___
*[pkg/roachpb/api.proto, line 2276 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/68115#-Mftb0YY8AQl0NA4yyRy:-Mftb0YY8AQl0NA4yyRz:bs75whm) ([raw file](https://github.com/cockroachdb/cockroach/blob/0f9241607580ce768750cfdd2d5c5c646e6f8dcb/pkg/roachpb/api.proto#L2276)):*
> ```protobuf
>   uint64 tenant_id = 2 [(gogoproto.customname) = "TenantID"];
> 
>   // ID of the SQL pod instance from where the request originates.
> ```

NIT: Would be nice to start with field name, using Go formatting, i.e. `InstanceID`.


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
TFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68115#-:-MfxG_m-B2qVz-JFnq4J:b-814g8k)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @adityamaru, @ajwerner, and @andy-kimball)

___
*[pkg/ccl/multitenantccl/tenantcostserver/metrics.go, line 112 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/68115#-Mft_uNSAVVC_NwmsZC2:-MfxF9btDY1pMGSVUgN_:b-g9t0dv) ([raw file](https://github.com/cockroachdb/cockroach/blob/0f9241607580ce768750cfdd2d5c5c646e6f8dcb/pkg/ccl/multitenantccl/tenantcostserver/metrics.go#L112)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

I don't think this needs to be a pointer, does it? Why allocate an extra object?
</blockquote></details>

We store this struct in the map and pass it around by value 
___
*[pkg/ccl/multitenantccl/tenantcostserver/system_table.go, line 33 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/68115#-MftaAH59cxD7DDdHgXq:-MfxFaDeBkbriPapayQj:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0f9241607580ce768750cfdd2d5c5c646e6f8dcb/pkg/ccl/multitenantccl/tenantcostserver/system_table.go#L33)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

SUPER NIT: These comments should start with field name to be idiomatic go; similar issues in `instanceState`.
</blockquote></details>

Done.
___
*[pkg/ccl/multitenantccl/tenantcostserver/token_bucket.go, line 82 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/68115#-Mfs6pYNDlnGaaq8oUCL:-MfxFl14AZKkW8c3MI3A:bj994qp) ([raw file](https://github.com/cockroachdb/cockroach/blob/a86f7187a3c2dfc36da3fec6550865aa4283d6dd/pkg/ccl/multitenantccl/tenantcostserver/token_bucket.go#L82)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Whoops, my comment got cut off. I meant to say, "This could be task for someone else". Also, you could add the work to fix the TODO in that task by adding an `UpdateIfHigher` or `RatchetUp` method. How come you removed the TODO comment from the metrics update code below?
</blockquote></details>

The mutex fixes that, the metric update is now atomic with the transaction.
___
*[pkg/roachpb/api.proto, line 2276 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/68115#-Mftb0YY8AQl0NA4yyRy:-MfxG0UF66r12lOiQWwe:b-hm1v9x) ([raw file](https://github.com/cockroachdb/cockroach/blob/0f9241607580ce768750cfdd2d5c5c646e6f8dcb/pkg/roachpb/api.proto#L2276)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

NIT: Would be nice to start with field name, using Go formatting, i.e. `InstanceID`.
</blockquote></details>

Done. I usually avoid that in .proto files because the name in the proto is not the same as the go name, but looking at the existing file, there is a mix of both practices.


<!-- Sent from Reviewable.io -->

andy-kimball reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68115)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @adityamaru, @ajwerner, @andy-kimball, and @RaduBerinde)

___
*[pkg/ccl/multitenantccl/tenantcostserver/metrics.go, line 112 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/68115#-Mft_uNSAVVC_NwmsZC2:-MfxwyKhCmjvdYTrSj0L:b5o8c9z) ([raw file](https://github.com/cockroachdb/cockroach/blob/0f9241607580ce768750cfdd2d5c5c646e6f8dcb/pkg/ccl/multitenantccl/tenantcostserver/metrics.go#L112)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

We store this struct in the map and pass it around by value 
</blockquote></details>

I see, OK.
___
*[pkg/ccl/multitenantccl/tenantcostserver/token_bucket.go, line 82 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/68115#-Mfs6pYNDlnGaaq8oUCL:-MfxwF1O4YI0m2tpLGrx:bkclf7z) ([raw file](https://github.com/cockroachdb/cockroach/blob/a86f7187a3c2dfc36da3fec6550865aa4283d6dd/pkg/ccl/multitenantccl/tenantcostserver/token_bucket.go#L82)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

The mutex fixes that, the metric update is now atomic with the transaction.
</blockquote></details>

Ah, got it.


<!-- Sent from Reviewable.io -->

RaduBerinde commented:
bors r+
